custom/ilsgateway/tanzania/reports/unrecognized_sms.py

```diff
@@ -0,0 +1,165 @@
+        )
+
+    def _fmt_contact_link(self, msg, doc_info):
+        if doc_info:
+            username, contact_type, url = (doc_info.display,
+                doc_info.type_display, doc_info.link)
+        else:
+            username, contact_type, url = (None, None, None)
+        username = username or "-"
+        contact_type = contact_type or _("Unknown")
+        if url:
+            ret = self.table_cell(username, '<a href="%s">%s</a>' % (url, username))
+        else:
+            ret = self.table_cell(username, username)
+        ret['raw'] = "|||".join([username, contact_type,
+            msg.couch_recipient or ""])
+        return ret
+
+    def get_recipient_info(self, message, contact_cache):
+        recipient_id = message.couch_recipient
+
+        if recipient_id in contact_cache:
+            return contact_cache[recipient_id]
+
+        doc = None
+        if recipient_id not in [None, ""]:
+            try:
+                if message.couch_recipient_doc_type == "CommCareCase":
+                    doc = CommCareCase.get(recipient_id)
+                else:
+                    doc = CouchUser.get_by_user_id(recipient_id)
+            except Exception:
+                pass
+
+        if doc:
+            doc_info = get_doc_info(doc.to_json(), self.domain)
+        else:
+            doc_info = None
+
+        contact_cache[recipient_id] = doc_info
+
+        return doc_info
+
+    @property
+    def export_table(self):
+        result = super(UnrecognizedSMSReport, self).export_table
+        table = result[0][1]
+        table[0].append(_("Contact Type"))
+        table[0].append(_("Contact Id"))
+        for row in table[1:]:
+            contact_info = row[1].split("|||")
+            row[1] = contact_info[0]
+            row.append(contact_info[1])
+            row.append(contact_info[2])
+        return result
+
+    def get_location_filter(self):
+        locations = []
+        location_id = AsyncLocationFilter.get_value(self.request, self.domain)
+        if location_id:
+            locations = SQLLocation.objects.get(
+                location_id=location_id
+            ).get_descendants(
+                include_self=True
+            ).filter(
+                location_type__administrative=False
+            ).values_list('location_id', flat=True)
+
+        return locations
+
+    @property
+    def headers(self):
+        header = DataTablesHeader(
+            DataTablesColumn(_("Timestamp")),
+            DataTablesColumn(_("User Name")),
+            DataTablesColumn(_("Phone Number")),
+            DataTablesColumn(_("Direction")),
+            DataTablesColumn(_("Message")),
+        )
+        header.custom_sort = [[0, 'desc']]
+        return header
+
+    @property
+    def rows(self):
+        startdate = json_format_datetime(self.datespan.startdate_utc)
+        enddate = json_format_datetime(self.datespan.enddate_utc)
+        data = SMSLog.by_domain_date(self.domain, startdate, enddate)
```

Comment:
are we going to paginate this at all? this query is going to be huge if you choose a large date range. cc @gcapalbo since I think he's looking at migrating these objects to SQL